src/multiset.rs

```diff
@@ -340,6 +340,239 @@ where
+            cur_count: 0,
+        }
+    }
+
+    /// Returns an iterator over the union of the two sets.
+    ///
+    /// This entails yielding each key with the maximum number of times it appears in either set.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use multiset::HashMultiSet;
+    ///
+    /// let set1: HashMultiSet<_> = [1, 1, 2].iter().cloned().collect();
+    /// let set2: HashMultiSet<_> = [2, 2, 3].iter().cloned().collect();
+    /// let set_union: HashMultiSet<_> = set1
+    ///     .union_counts(&set2)
+    ///     .map(|(&key, count)| (key, count))
+    ///     .collect();
+    /// assert_eq!(set_union.count_of(&1), 2);
+    /// assert_eq!(set_union.count_of(&2), 2);
+    /// assert_eq!(set_union.count_of(&3), 1);
+    /// ```
+    pub fn union_counts<'a>(&'a self, other: &'a HashMultiSet<K>) -> UnionCounts<'a, K> {
+        UnionCounts {
+            set1: self,
+            set2: other,
+            set1_iter: self.elem_counts.iter(),
+            set2_iter: other.elem_counts.iter(),
+        }
+    }
+
+    /// Return an iterator over the intersection of the two sets.
+    ///
+    /// This entails yielding each key with the minimum number of times it appears in either set. If
+    /// a key only appears in one set, it will not appear at all in the intersection.
+    ///
+    /// ```
+    /// use multiset::HashMultiSet;
+    ///
+    /// let set1: HashMultiSet<_> = [1, 1, 2].iter().cloned().collect();
+    /// let set2: HashMultiSet<_> = [2, 2, 3].iter().cloned().collect();
+    /// let set_intersection: HashMultiSet<_> = set1
+    ///     .intersection_counts(&set2)
+    ///     .map(|(&key, count)| (key, count))
+    ///     .collect();
+    /// assert_eq!(set_intersection.count_of(&1), 0);
+    /// assert_eq!(set_intersection.count_of(&2), 1);
+    /// assert_eq!(set_intersection.count_of(&3), 0);
+    /// ```
+    pub fn intersection_counts<'a>(
+        &'a self,
+        other: &'a HashMultiSet<K>,
+    ) -> IntersectionCounts<'a, K> {
+        IntersectionCounts {
+            set2: other,
+            set1_iter: self.elem_counts.iter(),
+        }
+    }
+}
+
+/// An iterator over the union of two structs. See [HashMultiSet::union] for
+/// more details.
+pub struct Union<'a, K> {
+    set1_iter: hash_map::Iter<'a, K, usize>,
+    set2_iter: hash_map::Iter<'a, K, usize>,
+    set1: &'a HashMultiSet<K>,
+    set2: &'a HashMultiSet<K>,
+    cur_entry: Option<(&'a K, usize)>,
+    cur_count: usize,
+}
+
+impl<'a, K> Iterator for Union<'a, K>
+where
+    K: Eq + Hash,
+{
+    type Item = &'a K;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some((key, count)) = self.cur_entry {
+            if self.cur_count < count {
+                self.cur_count += 1;
+                return Some(key);
+            }
+        }
+        if let Some((new_key, &new_count)) = self.set1_iter.next() {
+            let max_count = std::cmp::max(new_count, self.set2.count_of(new_key));
```

Comment:
As mentioned above this should probably be: `self.set2.count_of(new_key) + new_count`